src/basic/fs-util.c

```diff
@@ -930,3 +930,75 @@ int access_fd(int fd, int mode) {
+         * able to keep the data pinned on disk forever. This call is particular useful whenever we execute clean-up
+         * jobs ("vacuuming"), where we want to make sure the data is really gone and the disk space released and
+         * returned to the free pool.
+         *
+         * Deallocation is preferably done by FALLOC_FL_PUNCH_HOLE|FALLOC_FL_KEEP_SIZE (👊) if supported, which means
+         * the file won't change size. That's a good thing since we shouldn't needlessly trigger SIGBUS in other
+         * programs that have mmap()ed the file. (The assumption here is that changing file contents to all zeroes
+         * underneath those programs is the better choice than simply triggering SIGBUS in them which truncation does.)
+         * However if hole punching is not implemented in the kernel or file system we'll fall back to normal file
+         * truncation (🔪), as our goal of deallocating the data space trumps our goal of being nice to readers (💐).
+         *
+         * Note that we attempt deallocation, but failure to succeed with that is not considered fatal, as long as the
+         * primary job – to delete the file – is accomplished. */
+
+        if ((flags & AT_REMOVEDIR) == 0) {
+                truncate_fd = openat(fd, name, O_WRONLY|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW|O_NONBLOCK);
+                if (truncate_fd < 0) {
+
+                        /* If this failed because the file doesn't exist propagate the error right-away. Also,
+                         * AT_REMOVEDIR wasn't set, and we tried to open the file for writing, which means EISDIR is
+                         * returned when this is a directory but we are not supposed to delete those, hence propagate
+                         * the error right-away too. */
+                        if (IN_SET(errno, ENOENT, EISDIR))
+                                return -errno;
+
+                        log_debug_errno(errno, "Failed to open file '%s' for deallocation, ignoring: %m", name);
+                }
+        }
+
+        if (unlinkat(fd, name, flags) < 0)
+                return -errno;
+
+        if (truncate_fd < 0) /* Don't have a file handle, can't do more ☹️ */
+                return 0;
+
+        if (fstat(truncate_fd, &st) < 0) {
+                log_debug_errno(errno, "Failed to stat file '%s' for deallocation, ignoring.", name);
+                return 0;
+        }
+
+        if (!S_ISREG(st.st_mode) || st.st_blocks == 0 || st.st_nlink > 0)
+                return 0;
+
+        /* If this is a regular file, it actually took up space on disk and there are no other links it's time to
+         * punch-hole/truncate this to release the disk space. */
+
+        bs = MAX(st.st_blksize, 512);
+        l = DIV_ROUND_UP(st.st_size, bs) * bs; /* Round up to next block size */
+
+        if (fallocate(truncate_fd, FALLOC_FL_PUNCH_HOLE|FALLOC_FL_KEEP_SIZE, 0, l) >= 0)
```

Comment:
`systemd` seems to be failing to compile on `CentOS` with the following error:
```sh
FAILED: src/basic/basic@sta/fs-util.c.o
cc  -Isrc/basic/basic@sta -Isrc/basic -I../src/basic -I. -I../ -Isrc/libsystemd-network -I../src/libsystemd-network -I../src/libsystemd/sd-network -I../src/libsystemd/sd-netlink -I../src/libsystemd/sd-id128 -I../src/libsystemd/sd-hwdb -I../src/libsystemd/sd-device -I../src/libsystemd/sd-bus -Isrc/core -I../src/core -Isrc/libudev -I../src/libudev -Isrc/udev -I../src/udev -Isrc/login -I../src/login -Isrc/timesync -I../src/timesync -Isrc/resolve -I../src/resolve -Isrc/journal -I../src/journal -Isrc/systemd -I../src/systemd -Isrc/shared -I../src/shared -I/usr/include/blkid -I/usr/include/uuid -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=gnu99 -O0 -g -Wextra -Werror=undef -Wlogical-op -Wmissing-include-dirs -Wold-style-definition -Wpointer-arith -Winit-self -Wdeclaration-after-statement -Wfloat-equal -Wsuggest-attribute=noreturn -Werror=missing-prototypes -Werror=implicit-function-declaration -Werror=missing-declarations -Werror=return-type -Werror=format=2 -Wstrict-prototypes -Wredundant-decls -Wmissing-noreturn -Wshadow -Wendif-labels -Wstrict-aliasing=2 -Wwrite-strings -Werror=overflow -Wnested-externs -ffast-math -fno-common -fdiagnostics-show-option -fno-strict-aliasing -fvisibility=hidden -fstack-protector -fstack-protector-strong --param=ssp-buffer-size=4 -fPIE -Wno-unused-parameter -Wno-missing-field-initializers -Wno-unused-result -Wno-error=nonnull -Werror=shadow -include config.h -g -O0 -ftrapv -fPIC -pthread -fvisibility=default -MMD -MQ 'src/basic/basic@sta/fs-util.c.o' -MF 'src/basic/basic@sta/fs-util.c.o.d' -o 'src/basic/basic@sta/fs-util.c.o' -c ../src/basic/fs-util.c
../src/basic/fs-util.c: In function ‘unlinkat_deallocate’:
../src/basic/fs-util.c:990:36: error: ‘FALLOC_FL_PUNCH_HOLE’ undeclared (first use in this function)
         if (fallocate(truncate_fd, FALLOC_FL_PUNCH_HOLE|FALLOC_FL_KEEP_SIZE, 0, l) >= 0)
                                    ^
../src/basic/fs-util.c:990:36: note: each undeclared identifier is reported only once for each function it appears in
../src/basic/fs-util.c:990:57: error: ‘FALLOC_FL_KEEP_SIZE’ undeclared (first use in this function)
         if (fallocate(truncate_fd, FALLOC_FL_PUNCH_HOLE|FALLOC_FL_KEEP_SIZE, 0, l) >= 0)
                                                         ^
```
, which is most likely because https://sourceware.org/git/?p=glibc.git;a=commit;h=257dabc059447934a90be6fce4b5d2a is not included in the version of `glibc` being used there. I think that `missing.h` needs updating.